### PR TITLE
[Don't Merge] Issue 258 - Update to rank and fix operator

### LINKED ISSUE #258: [BUG] expect_column_most_common_value_to_be_in_set handling of ties
### Is this a new bug in dbt-expectations?

- [x] I believe this is a new bug in dbt-expectations
- [x] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When most common values are non-unique, running expect_column_most_common_value_to_be_in_set only works for 1 of the most common values.

### Expected Behavior

expect_column_most_common_value_to_be_in_set should match all most common values

### Steps To Reproduce

in dbt-expectations/integration_tests/models/schema_tests/schema.yml, add following column test for model data_test column col_string_b.:
``` yaml
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['ab']
              top_n: 1
```
Although all values are distinct in that model column (i.e. all column values are "the most common"), any value in value_set different from 'b' (the first result naturally ordered) will fail.

### Relevant log output

```sql
with value_counts as (
    select
        col_string_b as value_field,
        count(*) as value_count
    from dbt_expectations_integration_tests.data_test
    group by col_string_b
),
value_counts_ranked as (
    select
        *,
        row_number() over(order by value_count desc) as value_count_rank
    from
        value_counts
),
value_count_top_n as (
    select
        value_field
    from
        value_counts_ranked
    where
        value_count_rank = 1
),
set_values as (
    select
        'ab' as value_field
),
unique_set_values as (
    select distinct value_field
    from
        set_values
),
validation_errors as (
    -- values from the model that are not in the set
    select
        value_field
    from
        value_count_top_n
    where
        value_field not in (select value_field from unique_set_values)
)
select *
from validation_errors
```
### Environment

```markdown
- OS: Manjaro Linux 22.0.2
- Python: 3.10.9
- dbt: 1.5.0
- dbt-expectations: 0.8.5
```

### Which database adapter are you using with dbt?

Postgres

Note: dbt-expectations currently does not support database adapters other than the ones listed below.

- Postgres
- Snowflake
- BigQuery

### Additional Context

[Great expectations](https://greatexpectations.io/expectations/expect_column_most_common_value_to_be_in_set) has an additional ties_okay argument for partial matches. We could also add support for that.

So following tests would succeed
``` yaml
            # Expect error if not all most common values are in the set
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['b']
              top_n: 1
              config:
                error_if: "=0"
                warn_if: "<3"
            # Expect success if not all most common values are in the set but ties_okay is set
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['b']
              top_n: 1
              ties_okay: true
            # Expect error if none of the most common values are in the set and ties_okay is set
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['invalid_value']
              top_n: 1
              ties_okay: true
              config:
                error_if: "=0"
                warn_if: "<4"
            # Expect success if not all most common values are in the set but ties_okay is set
            # and the set contains extra values 
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['b', 'invalid_value']
              top_n: 1
              ties_okay: true
            # Expect success if not all most common values are in the set but ties_okay is set
            # and value is not first one of the column naturally ordered
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['ab']
              top_n: 1
              ties_okay: true
            # Expect success if all most common values are in the set
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['b', 'ab', 'abc', 'abcd']
              top_n: 1
            # Expect success if all most common values are in the set 
            # and the set contains extra values 
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['b', 'ab', 'abc', 'abcd', 'invalid_value']
              top_n: 1
            # Expect error if none of the most common values are in the set 
            # and the set contains extra values 
          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
              value_set: ['invalid_value1', 'invalid_value2', 'invalid_value3', 'invalid_value4', 'invalid_value5']
              top_n: 1
              config:
                error_if: "=0"
                warn_if: "<4"
```

### PR DESCRIPTION
**Pull request by clausherther**
## Issue this PR Addresses/Closes

Closes #258

## Summary of Changes
Simplified example to explain suggestion made [here](https://github.com/calogica/dbt-expectations/pull/259/files#r1198452316)

## Reviewers
@clausherther

### CODE REVIEW

File: integration_tests/models/schema_tests/schema.yml [hunk 1]
```diff
@@ -505,6 +505,20 @@ models:
               value_set: [0.5]
               top_n: 1
               quote_values: false
+              data_type: float
+          - dbt_expectations.expect_column_most_common_value_to_be_in_set:
+              value_set: [0.5, 0, 1]
+              top_n: 2
+              quote_values: false
+              data_type: float
```
Comment (clausherther) [inline integration_tests/models/schema_tests/schema.yml:513]: fyi, Snowflake won't correctly compare a `float` in the db to a value cast to `decimal` (default for the test)

File: macros/schema_tests/aggregate_functions/expect_column_most_common_value_to_be_in_set.sql [hunk 2]
```diff
@@ -22,6 +22,8 @@
                                                                       row_condition
                                                                       ) %}
 
+{% set data_type = dbt.type_numeric() if not data_type else data_type %}
```
Comment (clausherther) [inline macros/schema_tests/aggregate_functions/expect_column_most_common_value_to_be_in_set.sql:25]: fyi, Snowflake won't correctly compare a float in the db to a value cast to decimal (default for the test) so makes more sense to use the built-in types across the test to make sure comparisons actually work.

Comment (lookslikeitsnot) [reply macros/schema_tests/aggregate_functions/expect_column_most_common_value_to_be_in_set.sql:25]: From [the official documentation](https://docs.snowflake.com/en/sql-reference/data-types-numeric) and my testing,  inserting floating point data will create a `number` column type with the smallest precision and scale to fit the value. e.g.:  when creating a table with initial data, if a column contains [0, 1, 0.5], the assigned type will be `number(2,1)`. So no information is lost at that point. However casting that to "numeric", "decimal", "number" type without explicit precision and scale will result in rounding and casting it to any "float" or "double" type [has the usual accuracy problems](https://en.wikipedia.org/wiki/Floating-point_arithmetic#Accuracy_problems).

When comparing, Snowflake will happily attempt to compare any numeric type with the usual issues from floats
 <table>
  <tr>
    <th>comparison</th>
    <th>result</th>
  </tr>
  <tr>
   <td>cast(0.5 as number(2,1)) = cast(0.5 as numeric(28,2))</td>
   <td>TRUE</td>
     </tr>
  <tr>
   <td>cast(0.5 as float) = cast(0.5 as numeric(28,2))</td>
   <td>TRUE</td>
     </tr>
  <tr>
    <td>cast(0.5 as float) = cast(0.5 as numeric)</td>
    <td>FALSE</td>
     </tr>
<tr>
    <td>0.1 + 0.2 = 0.3</td>
    <td>TRUE</td>
  </tr>
<tr>
    <td>(cast(0.1 as float) + cast(0.2 as float)) = cast(0.3 as float)</td>
    <td>FALSE</td>
  </tr>
  <tr>
    <td>cast(3.1415926535897932384626433832795028841 as number(38,37)) = cast(3.141592653589793 as float)</td>
    <td>TRUE</td>
  </tr>
</table> 

*TL;DR*: It makes sense to use built-in types everywhere to not get db-engine specific issues.

File: integration_tests/models/schema_tests/data_test.sql [hunk 3]
```diff
@@ -13,8 +13,8 @@ union all
 select
     2 as idx,
     '2020-10-22' as date_col,
-    1 as col_numeric_a,
-    0 as col_numeric_b,
+    cast(1 as {{ dbt.type_numeric() }}) as col_numeric_a,
```
Comment (lookslikeitsnot) [inline integration_tests/models/schema_tests/data_test.sql:16]: Is there any database engine that wouldn't automatically cast every subsequent value on insert when the first value has an explicit type? (it doesn't hurt to cast them all, I'm curious and couldn't find the information online)
